src/mpid/ch4/src/ch4_impl.h

```diff
@@ -935,4 +935,89 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_CH4U_wait_am_acc(MPIR_Win * win, int target_r
   fn_fail:
     goto fn_exit;
 }
+
+/* Collectively allocate shared memory region.
+ * MPL_shm routines and MPI collectives are internally used.
+ *
+ * TODO: If symmetric_flag is true, then the routine will try to get
+ * a symmetric address and give up after tried predefined times. */
+#undef FUNCNAME
+#define FUNCNAME MPIDI_CH4U_allocate_shm_segment
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+static inline int MPIDI_CH4U_allocate_shm_segment(MPIR_Comm * shm_comm_ptr,
+                                                  MPI_Aint shm_segment_len, int symmetric_flag,
+                                                  MPL_shm_hnd_t * shm_segment_hdl_ptr,
+                                                  void **base_ptr)
+{
+    MPIR_Errflag_t errflag = MPIR_ERR_NONE;
+    int mpi_errno = MPI_SUCCESS;
+    MPL_shm_hnd_t shm_segment_handle;
+    void *map_ptr = NULL;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_CH4R_ALLOCATE_SHM_SEGMENT);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_CH4R_ALLOCATE_SHM_SEGMENT);
+
+    mpi_errno = MPL_shm_hnd_init(&shm_segment_handle);
+    if (mpi_errno)
+        MPIR_ERR_POP(mpi_errno);
+
+    if (shm_comm_ptr->rank == 0) {
+        char *serialized_hnd_ptr = NULL;
+
+        /* create shared memory region for all processes in win and map */
+        mpi_errno = MPL_shm_seg_create_and_attach(shm_segment_handle,
+                                                  shm_segment_len, (char **) &map_ptr, 0);
+        if (mpi_errno)
+            MPIR_ERR_POP(mpi_errno);
+
+        /* serialize handle and broadcast it to the other processes in win */
+        mpi_errno = MPL_shm_hnd_get_serialized_by_ref(shm_segment_handle, &serialized_hnd_ptr);
+        if (mpi_errno)
+            MPIR_ERR_POP(mpi_errno);
+
+        mpi_errno = MPIR_Bcast(serialized_hnd_ptr, MPL_SHM_GHND_SZ, MPI_CHAR, 0,
+                               shm_comm_ptr, &errflag);
+        MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+
+        /* wait for other processes to attach to win */
+        mpi_errno = MPIR_Barrier(shm_comm_ptr, &errflag);
+        MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+
+        /* unlink shared memory region so it gets deleted when all processes exit */
+        mpi_errno = MPL_shm_seg_remove(shm_segment_handle);
```

Comment:
Question: why shm handle is removed on rank 0 only?

Reply:
Emm, `MPL_shm_seg_remove` internally calls `unlink`. `unlink` deletes the file from the file system once all accessing processes call `close` on the file (i.e., `MPIDI_CH4R_win_finalize` -> `MPL_shm_hnd_finalize` -> `close`). Thus, I think we should only call `MPL_shm_seg_remove` on rank 0 here.

Reply:
Thanks for explanation.